### DISCUSSION TITLE
Developer Set Sign-in and Jr-Developers

### DISCUSSION THREAD

**Discussion by kdecks** [RetroAchievements/RAWeb#706] [Ideas]
### Preface
*Jump to the bottom for benefits. This is all very much tied into #57.

#### Signing in
1. Done in the ``dev`` tools on the game page.
2. It allows the developer to add new achievements to official. *Not being signed in would prevent it.*
3. The set will automatically **Progress Flag**ged (see #57) as ``🔨 Work-in-progress`` (aka ``WIP``)  by 
4. A report of all sets signed into and by whom would be made available to the community.
[Developer]"
#### Limits
1. ``Developers`` would have a limit of 3 ``WIP`` sets,
2. ``Jr-Developers`` would have a limit of 1 ``WIP`` set.
3. Sets ``Complete`` (see my first post in #57) would be **locked** from development until they had been authorized to add new achievements though the approval process. (i.e. after presenting a plan they would be approved automatically or a vote would take place, if there were objections. An admin would need to manually unlock the set.)
4. Cap by default the number of developers that sign in to one. A developer who is signed in can increase the cap for cooperative work with others.  (This prevents a developer from making achievements "out from under" another's work; (this has happened recently and occasionally in the past.))
5. If a set is **locked** a developer will not be able to sign-in and will be redirected to how to get permission for a revision.
6. If a set is Signed in and no new achievements have been added within 1 month they are kicked out automatically, and the **Progress Flag** is set from ``WIP`` to ``Empty`` or ``Incomplete`` depending on if there are achievements. If kicked a developer can sign back in.
#### Permissions
1. Admin: Add/Remove developers from a set.
2. Admin: **Lock/unlock** a set, A set is normally locked when a developer signs out as ``Complete`` (see #57 ).
3. Developer (who is signed in): Increase cap of developer that can sign into the set.
---

``Jr-Developers`` only have value if a the Developer sign-in concept was put in place (see above).

### About Jr-Developers
- ``Jr-Developers`` would be given the status on request, and promoted to ``Developer`` after completion of 1-3 sets (leaning on 2). 
   - **ALT** ``Jr-Developers`` would be required to work in tandem with a ``Developer`` on a set from start to finish before gaining full ``Developer`` status. *This seems like a nice concept, but likely would require too much coordination; I still want to mention it.*
- ``Jr-Developers`` would be limited to working on 1 set. (see Sign-in concept below)
- (optional) ``Jr-Developers`` would have achievement tools visible. (Currently not restricted, but there are arguments for it being restricted in the future.)
- ~~(optional) ``Jr-Developers`` would lose jr-dev status automatically if no achievements are published after 3 months.~~
----

### Benefits
1. Self-regulating. No rules to know or follow.
2. At any time a developer will know if they are free to work on a set.
3. The community will know the status of a given set.
4. The community will know which developers are working on what.
5. Developers will be limited in the number of sets they are working on, opening games that are supposedly work-in-progress but not progressing to the community.
6. Mostly automatic, minus some needed changes by admin (see **Permissions** above).
7. Aspiring and current Developers are given more defined structure.
8. Transparency.

**Comment by SyrianBallaS** [comment]
I think instead of 3 month lockout, jr dev should be required to write a
status report.

 I say that since, to my knowledge, this isn't a for profit org where we
have ship products for clients.

**Comment by kdecks** [comment]
Developers and Jr-developers would not be locked out of working on sets, but limited to the number they can work on at once. 1 and 3(3-6) respectively. There's no lock-out per se at all, except for going over that limit. There is a time-out, as in if you've logged into a set and didn't do anything you'd be kicked after a period of inactivity (a month). 

edit: Oh, I see what you mean. Yeah the idea is if a jr-dev requests to be a dev and then doesn't do anything with it they lose the status, which could be requested again. I've revised the wording.

**Comment by julkasalsa** [comment]
How about a limit for one set per system instead (to promote less popular systems this way)? Additional there could be 2 sets on one system at the same time (but only for one system). For regular devs I mean. This way someone who would like to switch between a lot of games will still have good variety.

**Comment by ikki5** [comment]
> (optional) Jr-Developers would lose jr-dev status automatically if no achievements are published after 3 months.

For this, I would say more status updates are required or if status updates are still going and on going, then the 3 month rule would be omitted. I look at Flara's case where she was starting on one game and then she moved to another, not only that, she was custom making all her badges which I am pretty sure she spend like a month of that alone (along with real life work and that as well).

> Jr-Developers would be given the status on request, and promoted to Developer after completion of 1-3 sets (leaning on 2).
> 
>     ALT Jr-Developers would be required to work in tandem with a Developer on a set from start to finish before gaining full Developer status. This seems like a nice concept, but likely would require too much coordination; I still want to mention it.

This, I do like this idea, give them 1-3 sets and probably an average of 2. And as for the "Alt," i think that working with an already established developer, or at least least getting active help from a developer would maybe get them to only needing a single set (in most cases) to make the developer. And no, this would be a bit more than just asking for help in the dev board, but more one of the devs kind of being a "mentor." it does not mean the current dev to actually need to code things or whatever, but pretty much be there, discuss stuff with them, help them out, give tips and such. This would also allow less co-ordination that needs to be involved as the jr. dev can work away and still get help, input and advice and criticism for the achievements being made and/or ideas and such.

Now, as for the limits, as long as you don't have to be signed into a game/set to be able to work on it 9talking about games that are not WIP for anyone) to be able to do stuff with them, from discord, this sounded more like laying a claim on the games rather than being told you can only work on "3 sets."

**Comment by kdecks** [comment]
Yeah, losing Jr. dev status in this way is no good. Scratch that idea.

That mentorship stuff is really good. If we could get some flow there I think it could be really positive.

You'd not be required to sign into a game to look at the memory, add code notes or create achievements in local. You would need to be signed in to add achievements to unofficial or official.

**Comment by Jamiras** [comment]
This is largely implemented by #1026. I'm going to leave it open as I'm not sure that it's been fully addressed.
